rust/types/src/partitioned_mutex.rs

```diff
@@ -0,0 +1,60 @@
+use std::{
+    hash::{DefaultHasher, Hash, Hasher},
+    marker::PhantomData,
+    sync::Arc,
+};
+
+#[derive(Clone)]
+pub struct AysncPartitionedMutex<K, V = (), H = DefaultHasher>
+where
+    K: Hash + Eq,
+    H: Hasher + Default,
+    V: Clone,
+{
+    partitions: Arc<[tokio::sync::Mutex<V>]>,
+    _hasher: std::marker::PhantomData<H>,
+    _key: std::marker::PhantomData<K>,
+}
+
+// TODO: A sensible value for this.
+const DEFAULT_NUM_PARTITIONS: usize = 16 * 16;
+
+impl<K, V, H> AysncPartitionedMutex<K, V, H>
+where
+    K: Hash + Eq,
+    H: Hasher + Default,
+    V: Clone,
+{
+    pub fn new(default_value: V) -> Self {
+        let partitions = (0..DEFAULT_NUM_PARTITIONS)
+            .map(|_| tokio::sync::Mutex::new(default_value.clone()))
+            .collect::<Vec<_>>();
+        Self {
+            partitions: partitions.into(),
+            _hasher: PhantomData,
+            _key: PhantomData,
+        }
+    }
+
+    fn with_partitions(num_partitions: usize, default_value: V) -> Self {
+        let partitions = (0..num_partitions)
+            .map(|_| tokio::sync::Mutex::new(default_value.clone()))
+            .collect::<Vec<_>>();
+        Self {
+            partitions: partitions.into(),
+            _hasher: PhantomData,
+            _key: PhantomData,
+        }
+    }
+
+    pub fn with_workers(num_workers: usize, default_value: V) -> Self {
```

Comment:
I am confused -finding the word workers very confusing- what is the point of this? Why not just let someone configure the partition count directly. Feel like im missing some knowledge

Reply:
update - addressed elsewhere. Birthday paradox says we make collisions less likely by doing square. We should just comment this

Reply:
changed the name and added comments